A few bugs related to static parameters

Three variations of a bug related to static parameters (doesn't matter if proc is annotated with `{.compileTime.}` or not.

### Variation 1
```nim
import unicode

type Cell = object
  c: Rune

proc test(c: static string) = discard #Remove this and it compiles
proc test(c: Cell) = discard

test Cell(c: runeAt(" ", 0))
```
fails with:
```
Hint: system [Processing]
Hint: widestrs [Processing]
Hint: io [Processing]
Hint: muscii [Processing]
Hint: unicode [Processing]
/home/clyybber/builds/nim/compiler/nim.nim(106) nim
/home/clyybber/builds/nim/compiler/nim.nim(83) handleCmdLine
/home/clyybber/builds/nim/compiler/cmdlinehelper.nim(98) loadConfigsAndRunMainCommand
/home/clyybber/builds/nim/compiler/main.nim(188) mainCommand
/home/clyybber/builds/nim/compiler/main.nim(92) commandCompileToC
/home/clyybber/builds/nim/compiler/modules.nim(144) compileProject
/home/clyybber/builds/nim/compiler/modules.nim(85) compileModule
/home/clyybber/builds/nim/compiler/passes.nim(210) processModule
/home/clyybber/builds/nim/compiler/passes.nim(86) processTopLevelStmt
/home/clyybber/builds/nim/compiler/cgen.nim(1871) myProcess
/home/clyybber/builds/nim/compiler/ccgstmts.nim(1263) genStmts
/home/clyybber/builds/nim/compiler/ccgexprs.nim(2584) expr
/home/clyybber/builds/nim/compiler/ccgcalls.nim(572) genCall
/home/clyybber/builds/nim/compiler/ccgcalls.nim(194) genPrefixCall
/home/clyybber/builds/nim/compiler/ccgcalls.nim(177) genArg
/home/clyybber/builds/nim/compiler/cgen.nim(593) initLocExprSingleUse
/home/clyybber/builds/nim/compiler/ccgexprs.nim(2610) expr
/home/clyybber/builds/nim/compiler/ccgexprs.nim(1368) genObjConstr
/home/clyybber/builds/nim/compiler/ccgexprs.nim(768) lookupFieldAgain
/home/clyybber/builds/nim/lib/system/assertions.nim(27) failedAssertImpl
/home/clyybber/builds/nim/lib/system/assertions.nim(20) raiseAssert
/home/clyybber/builds/nim/lib/system/fatal.nim(39) sysFatal
Error: unhandled exception: /home/clyybber/builds/nim/compiler/ccgexprs.nim(768, 11) `ty.kind in {tyTuple, tyObject}`  [AssertionError]
```

### Variation 2
```nim
type Cell = object
  c: int

proc test(c: static string) = discard #Remove this and it compiles
proc test(c: Cell) = discard

test Cell(c: 0)
```
fails with:
```
Hint: system [Processing]
Hint: widestrs [Processing]
Hint: io [Processing]
Hint: muscii [Processing]
/home/clyybber/projects/tests/symbolbind/muscii.nim(7, 10) Error: cannot create null element for: (c: 0)
```

### Variation 3
```nim
type Cell = object
  c: int

proc test(c: static string) = discard #Remove this and it compiles
proc test(c: Cell) = discard

test Cell()
```
fails with:
```
Hint: system [Processing]
Hint: widestrs [Processing]
Hint: io [Processing]
Hint: muscii [Processing]
/home/clyybber/projects/tests/symbolbind/muscii.nim(4, 6) Hint: 'test' is declared but not used [XDeclaredButNotUsed]
CC: muscii.nim
Error: execution of an external compiler program 'gcc -c  -w   -I/home/clyybber/builds/nim/lib -I/home/clyybber/projects/tests/symbolbind -o /home/clyybber/.cache/nim/muscii_d/@mmuscii.nim.c.o /home/clyybber/.cache/nim/muscii_d/@mmuscii.nim.c' failed with exit code: 1

/home/clyybber/.cache/nim/muscii_d/@mmuscii.nim.c: In function 'NimMainModule':
/home/clyybber/.cache/nim/muscii_d/@mmuscii.nim.c:125:2: error: conversion to non-scalar type requested
  125 |  T1_ = (tyObject_Cell__kmBW6cD01IjfZ6f5JDLCcg)0;
      |  ^~~
/home/clyybber/.cache/nim/muscii_d/@mmuscii.nim.c:126:2: error: conversion to non-scalar type requested
  126 |  T1_ = (tyObject_Cell__kmBW6cD01IjfZ6f5JDLCcg)0;
      |  ^~~
/home/clyybber/.cache/nim/muscii_d/@mmuscii.nim.c:127:32: error: incompatible type for argument 1 of 'test__Kx3WzCWNc9cOf2APfVplmdA'
  127 |  test__Kx3WzCWNc9cOf2APfVplmdA(T1_);
      |                                ^~~
      |                                |
      |                                tyObject_Cell__kmBW6cD01IjfZ6f5JDLCcg {aka struct tyObject_Cell__kmBW6cD01IjfZ6f5JDLCcg}
/home/clyybber/.cache/nim/muscii_d/@mmuscii.nim.c:79:101: note: expected 'tyObject_Cell__beQ9ceLyDS5Mtemd0Wyf0Qw' {aka 'struct tyObject_Cell__beQ9ceLyDS5Mtemd0Wyf0Qw'} but argument is of type 'tyObject_Cell__kmBW6cD01IjfZ6f5JDLCcg' {aka 'struct tyObject_Cell__kmBW6cD01IjfZ6f5JDLCcg'}
   79 | N_LIB_PRIVATE N_NIMCALL(void, test__Kx3WzCWNc9cOf2APfVplmdA)(tyObject_Cell__beQ9ceLyDS5Mtemd0Wyf0Qw c) { nimfr_("test", "/home/clyybber/projects/tests/symbolbind/muscii.nim");
      |                                                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
```




### Expected Output
All three should compile I guess?

### Additional Information

```
$ nim -v
Nim Compiler Version 1.1.1 (#devel)
```

--- a/compiler/sigmatch.nim
+++ b/compiler/sigmatch.nim
@@ -2004,6 +2004,7 @@ proc paramTypesMatchAux(m: var TCandidate, f, a: PType,
         let typ = newTypeS(tyStatic, c)
         typ.sons = @[evaluated.typ]
         typ.n = evaluated
+        arg = copyTree(arg) # fix #12864
         arg.typ = typ
         a = typ
       else:

--- a/lib/pure/typetraits.nim
+++ b/lib/pure/typetraits.nim
@@ -93,7 +93,7 @@ since (1, 1):
     # Note: `[]` currently gives: `Error: no generic parameters allowed for ...`
     type(default(T)[i])
 
-  type StaticParam*[value] = object
+  type StaticParam*[value: static type] = object
     ## used to wrap a static value in `genericParams`
 
 # NOTE: See https://github.com/nim-lang/Nim/issues/13758 - `import std/macros` does not work on OpenBSD
@@ -118,13 +118,12 @@ macro genericParamsImpl(T: typedesc): untyped =
           let ai = impl[i]
           var ret: NimNode
           case ai.typeKind
-          of ntyStatic:
-            since (1, 1):
-              ret = newTree(nnkBracketExpr, @[bindSym"StaticParam", ai])
           of ntyTypeDesc:
             ret = ai
+          of ntyStatic: doAssert false
           else:
-            assert false, $(ai.typeKind, ai.kind)
+            since (1, 1):
+              ret = newTree(nnkBracketExpr, @[bindSym"StaticParam", ai])
           result.add ret
         break
       else:

--- a/tests/metatype/tstaticparams.nim
+++ b/tests/metatype/tstaticparams.nim
@@ -171,3 +171,27 @@ echo inSize([
   [1, 2, 3],
   [4, 5, 6]
 ])
+
+block: # #12864
+  template fun() =
+    type Object = object
+    proc fun(f: Object): int = 1
+    proc fun(f: static[int]): int = 2
+    doAssert fun(Object()) == 1
+
+    var a: Object
+    doAssert fun(a) == 1
+
+    proc fun2(f: Object): int = 1
+    proc fun2(f: static[Object]): int = 2
+    doAssert fun2(Object()) == 2
+    doAssert fun2(a) == 1
+    const a2 = Object()
+    doAssert fun2(a2) == 2
+
+  fun()
+  static: fun()
+
+when true: #12864 original snippet
+  import times
+  discard times.format(initDateTime(30, mMar, 2017, 0, 0, 0, 0, utc()), TimeFormat())

--- a/tests/metatype/ttypetraits.nim
+++ b/tests/metatype/ttypetraits.nim
@@ -145,6 +145,7 @@ block genericParams:
   doAssert genericParams(Bar3).get(0) is StaticParam
   doAssert genericParams(Bar3).get(0).value == 3
   doAssert genericParams(Bar[3, float]).get(0).value == 3
+  static: doAssert genericParams(Bar[3, float]).get(0).value == 3
 
   type
     VectorElementType = SomeNumber | bool

--- a/tests/statictypes/t9255.nim
+++ b/tests/statictypes/t9255.nim
@@ -1,6 +1,6 @@
 discard """
   errormsg: '''
-type mismatch: got <static[proc (a0: int): string{.noSideEffect, gcsafe, locks: 0.}](bar)>
+type mismatch: got <proc (a0: int): string{.noSideEffect, gcsafe, locks: 0.}>
 '''
   line: 13
 """

--- a/tests/statictypes/tstatictypes.nim
+++ b/tests/statictypes/tstatictypes.nim
@@ -202,3 +202,17 @@ block: #13529
     proc fun[T](a: static[T]) = (const a2 = a)
     fun(1)
     fun(1.2)
+
+block: # #12713
+  block:
+    type Cell = object
+      c: int
+    proc test(c: static string) = discard #Remove this and it compiles
+    proc test(c: Cell) = discard
+    test Cell(c: 0)
+  block:
+    type Cell = object
+      c: int
+    proc test(c: static string) = discard #Remove this and it compiles
+    proc test(c: Cell) = discard
+    test Cell()

--- a/tests/statictypes/tstatictypes.nim
+++ b/tests/statictypes/tstatictypes.nim
@@ -170,3 +170,35 @@ var
   s: StringValue16
 
 echo s
+
+block: #13529
+  block:
+    type Foo[T: static type] = object
+    var foo: Foo["test"]
+    doAssert $foo == "()"
+    doAssert foo.T is string
+    static: doAssert foo.T == "test"
+    doAssert not compiles(
+      block:
+        type Foo2[T: static type] = object
+          x: T)
+
+  block:
+    type Foo[T: static[float]] = object
+    var foo: Foo[1.2]
+    doAssert $foo == "()"
+    doAssert foo.T == 1.2
+
+  block: # routines also work
+    proc fun(a: static) = (const a2 = a)
+    fun(1)
+    fun(1.2)
+  block: # routines also work
+    proc fun(a: static type) = (const a2 = a)
+    fun(1)
+    fun(1.2)
+
+  block: # this also works
+    proc fun[T](a: static[T]) = (const a2 = a)
+    fun(1)
+    fun(1.2)

